### ISSUE TITLE
custom gradient for int8

### ISSUE THREAD

**Issue by rybakov** [pytorch/pytorch#129889]
### 🚀 The feature, motivation and pitch

**Feature motivation:**
[Default pyTorch quantization aware training](https://pytorch.org/docs/stable/quantization.html) uses "fake-quantization" approach. For example for int8 quantization we will use float type during training and int8 during inference. There several disadvantages in this approach:
* It does not follow approach of "what you train is what your serve", as a result there can be numerical difference between training and inference (it emulates int ops using float ops and as soon as float summation goes beyond of 23 bits of float mantissa there will be a numerical difference after converting such model to int8).
* During training, it uses float operations, but it can use int8 operations to speed up training too.
* Current fake quantization aware training api uses both [python](https://github.com/pytorch/pytorch/tree/bb0f3df56279dff6782767804caebe5f6ffb2439/torch/ao/quantization) and [c++](https://github.com/pytorch/pytorch/tree/bb0f3df56279dff6782767804caebe5f6ffb2439/torch/csrc/jit/passes/quantization) levels, which makes it harder to support and extend. Instead if we use native pytorch types (e.g. int8, fp8) we could simplify quantization lib design and specify model quantization all in python. For example fp8 type is already used in [float8_experimental](https://github.com/pytorch-labs/float8_experimental).

To train a model with native int8 type I need two things:
* Control over output type in a matmul, e.g. for input int8 type return int32, as described in the [issues/128283](https://github.com/pytorch/pytorch/issues/128283)
* Custom gradient over int8 casting.

**Current issue:**
int8 tensors does nor have a gradient as discussed in ([link1](https://discuss.pytorch.org/t/is-the-type-conversion-differentiable/45436); [link2](https://discuss.pytorch.org/t/keep-the-gradient-function-when-changing-the-data-type-of-tensor/161639)). But it would be great if customer can specify their own gradient for int types in a straight  through estimator(STE).

There are two options to to do straight  through estimation (as shown below):
* Use custom backward (ste_custom_bwd).
* Use detach (ste_detach).

```
import torch

def ste_custom_bwd(func):
    """Straight through estimator based on custom backward."""
    class StraightThroughEstimator(torch.autograd.Function):
        @staticmethod
        def forward(ctx, input):
            return func(input)
        
        @staticmethod
        def backward(ctx, grad_output):
            return grad_output

    return StraightThroughEstimator.apply

def ste_detach(input, func):
  """Straight through estimator based on detach."""
  return input - input.detach() + func(input).detach()

def cast_int8(x):
    return x.to(torch.int8)
```

STE based on detach works ok:
```
x = torch.tensor([2.0], requires_grad=True)
y = ste_detach(x, cast_int8)
y.backward()
print("grad" + str(x.grad))
```

But STE based on custom backward pass, fails with expected message "element 0 of tensors does not require grad and does not have a grad_fn". 
```
x = torch.tensor([2.0], requires_grad=True)
y = ste_custom_bwd(cast_int8)(x)
y.backward()
print("grad" + str(x.grad))
```

It is not a blocker, because STE based on detach works, but it would be great if customer can specify custom backward pass for any type, for consistency with the API.

**Comment by vadimkantorov** [comment]
On the UX side, maybe in general an STE helper could be introduced to torch.higher_order ops or as some sort of pseudo-function (like proposed in https://github.com/pytorch/pytorch/issues/50122)

**Comment by HDCharles** [comment]
Perhaps @andrewor14 can comment by my understanding is as follows:

> It does not follow approach of "what you train is what your serve"....
pytorch quantization is for cpu but qat needs to be run on gpu for performance. The cpu kernels are fundamentally incompatible with this approach. Further the kernels cannot just separate out the matmul and quant/dequant operations, its all one kernel (at least for cpu) and we tend to find numerical differences even between the 'True' cpu/gpu quantized kernels.

>During training, it uses float operations, but it can use int8 operations to speed up training too.
I think we've looked into this, not sure what the result was, perhaps @andrewor14 can comment here

>Current fake quantization aware training api uses both [python](https://github.com/pytorch/pytorch/tree/bb0f3df56279dff6782767804caebe5f6ffb2439/torch/ao/quantization) and [c++](https://github.com/pytorch/pytorch/tree/bb0f3df56279dff6782767804caebe5f6ffb2439/torch/csrc/jit/passes/quantization) levels....[float8_experimental](https://github.com/pytorch-labs/float8_experimental)

float 8 experimental is being moved to [torchAO](https://github.com/pytorch/ao) where the various quantization techniques are developed for gpu. The python only code gets torch.compiled (which has its own difficulties) into triton code (not python). Its a very different set of developmental constraints compared to getting something to run fast on cpu which does tend to require c++

**Comment by yyfcc17** [comment]
```python
import torch
from torch.autograd import Function

class FloatToInt8(Function):
    @staticmethod
    def forward(ctx, input):
        """
        forward, quant float to int8, and record scale
        """
        scale = input.abs().max() / 127
        int8_output = (input / scale).clamp(-128, 127).round().to(torch.int8)
        ctx.save_for_backward(input, scale)
        return int8_output, scale

    @staticmethod
    def backward(ctx, grad_int8, grad_scale):
        """
        backward, use ste
        """
        print("grad_int8:", grad_int8)
        input, scale = ctx.saved_tensors
        grad_input = grad_int8.float() * (1.0 / scale)
        grad_input[input.abs() > (127 * scale)] = 0 
        return grad_input


class Int8ToFloat(Function):
    @staticmethod
    def forward(ctx, input, scale):
        """
        forward, dequant int8 to float
        """
        ctx.save_for_backward(scale, input)
        float_output = input.float() * scale
        return float_output

    @staticmethod
    def backward(ctx, grad_output):
        """
        backward, adjust grad with scale
        """
        scale, input = ctx.saved_tensors
        grad_input = grad_output * scale
        print("grad_input:", grad_input, "scale:", scale, "input:", input)
        grad_scale = None
        return grad_input, grad_scale


class TestPipeline(torch.nn.Module):
    def __init__(self):
        super(TestPipeline, self).__init__()

    def forward(self, input):
        int8_output, scale = FloatToInt8.apply(input)
        float_output = Int8ToFloat.apply(int8_output, scale)
        return float_output


if __name__ == "__main__":

    input = torch.tensor([1.5, -2.3, 3.9, -4.1], dtype=torch.float32, requires_grad=True)

    model = TestPipeline()

    output = model(input)
    print("Output:\n", output)

    loss = output.abs().sum()
    print("Loss:\n", loss)

    loss.backward()
    print("Input gradient:\n", input.grad)

```
the problem is, no matter what i give to `grad_input` in `Int8ToFloat` backward, the `grad_int8` received in `FloatToInt8` backward is always `zeros`.

a proposal:

in custom grad function, support any type and any shape of grad, no matter what the type of input tensor is, that is the meaning to write a `custom` grad function. what do you think?
